### PR TITLE
PML-370: merging NoisySLOS into release/0.4 (conflicts resolution)

### PR DESCRIPTION
**Pull request by CassNot**
## Summary

Merge the NoisySLOS work into `release/0.4` while preserving the contracts from the conflict resolution (hopefully):

- keep the NoisySLOS / NoisyG2SLOS execution contract
- keep the memristor state contract through `compute_with_keys`
- keep the new FidelityKernel backend contract
- keep the release/0.4 deprecation behavior

NB: complete noise support verification and demo for the Kernel will be handled in another ticket, along with the verification of
- NoiseModel with memristor
- NoiseModel with QGAN
- NoiseModel with QORC
- NoiseModel with QCNN when applicable
- NoiseModel with encoding space

## Related Issue
   
PML-370 

## Type of change
- [ ] Bug fix
- [x] New feature
- [x] Documentation update
- [ ] Refactor / Cleanup
- [ ] Performance improvement
- [ ] CI / Build / Tooling
- [ ] Breaking change (requires migration notes)

### NoisySLOS and noise routing

- Adds source-noise SLOS execution for indistinguishability and g2 noise.
- Adds `SectoredDistribution` support for g2/noisy source paths.
- Keeps public `QuantumLayer(... MeasurementStrategy.probs ...)` output as a `torch.Tensor` by default, even when g2 uses sectored internals.
- Preserves source, circuit, and post-measurement noise classification:
  - source: `indistinguishability`, `g2`, `g2_distinguishable`
  - circuit: `phase_imprecision`, `phase_error`
  - post-measurement: `brightness`, `transmittance`

### Memristor contract

- Restores `compute_with_keys(..., memristive_current_state=...)`.
- Threads the current memristive state through phase-error probability computation and circuit conversion.
- Keeps the memristive forward/update path compatible with keyed computations.

### Kernel backend contract

- Adapts `_CCInvQuantumLayer` / `FidelityKernel` to g2 sectored raw keys.
- Accepts `SectoredDistribution` from NoisyG2SLOS internally.
- Flattens/cleans sectored distributions before returning public tensor kernel outputs.
- Adds regressions for g2 kernel construction and evaluation.

### Deprecations

- Preserves release/0.4 deprecation behavior.
- Keeps `no_bunching` deprecation/error behavior.
- Makes the TorchScript `output_map_func` deprecation warning deterministic instead of relying on PyTorch version-specific warnings.

## Documentation

- Adds noisy simulation documentation in the user guide and quantum expert area.
- Adds API documentation entries for the new NoisySLOS and sectored distribution surface.
- Adds diagrams/images for noisy simulation explanations.

## How to test / How to run

1. Command lines

```
pytest tests/
```

## Documentation
- [x] User docs updated (Sphinx)
- [x] Examples / notebooks updated
- [x] Docstrings updated
- [x] Updated the API

## Checklist
- [x] PR title includes Jira issue key (e.g., PML-126)
- [x] "Related Jira ticket" section includes the Jira issue key (no URL)
- [x] Code formatted (ruff format)
- [x] Lint passes (ruff)
- [x] Static typing passes (mypy) if applicable
- [x] Unit tests added/updated (pytest)
- [x] Tests pass locally (pytest)
- [x] Tests pass on GPU (pytest)
- [x] Test coverage not decreased significantly
- [x] Docs build locally if affected (sphinx)
- [x] With this command: `SPHINXOPTS="-W --keep-going -n" make -C docs clean html` the docs are built without any warning or errors.
- [x] New public classes/methods/packages are added in the API following the methodology presented in other files.
- [x] Dependencies updated (if needed) and pinned appropriately
- [x] PR description explains what changed and how to validate it

### CODE REVIEW

**Comment by CassNot** [comment]
@claude review conflict merging

**Comment by CassNot** [comment]
@claude review Noisy SLOS support (indistinguishability, g2 and phase noise)

**Comment by ben9871** [comment]
> ## Summary
> Merge the NoisySLOS work into `release/0.4` while preserving the contracts from the conflict resolution (hopefully):
> 
> * keep the NoisySLOS / NoisyG2SLOS execution contract
> * keep the memristor state contract through `compute_with_keys`
> * keep the new FidelityKernel backend contract
> * keep the release/0.4 deprecation behavior
> 
> NB: complete noise support verification and demo for the Kernel will be handled in another ticket, along with the verification of
> 
> * NoiseModel with memristor
> * NoiseModel with QGAN
> * NoiseModel with QORC
> * NoiseModel with QCNN when applicable
> * NoiseModel with encoding space
> 
> ## Related Issue
> PML-370
> 
> ## Type of change
> * [ ]  Bug fix
> * [x]  New feature
> * [x]  Documentation update
> * [ ]  Refactor / Cleanup
> * [ ]  Performance improvement
> * [ ]  CI / Build / Tooling
> * [ ]  Breaking change (requires migration notes)
> 
> ## Proposed changes
> ### NoisySLOS and noise routing
> * Adds source-noise SLOS execution for indistinguishability and g2 noise.
> * Adds `SectoredDistribution` support for g2/noisy source paths.
> * Keeps public `QuantumLayer(... MeasurementStrategy.probs ...)` output as a `torch.Tensor` by default, even when g2 uses sectored internals.
> * Preserves source, circuit, and post-measurement noise classification:
>   
>   * source: `indistinguishability`, `g2`, `g2_distinguishable`
>   * circuit: `phase_imprecision`, `phase_error`
>   * post-measurement: `brightness`, `transmittance`
> 
> ### Memristor contract
> * Restores `compute_with_keys(..., memristive_current_state=...)`.
> * Threads the current memristive state through phase-error probability computation and circuit conversion.
> * Keeps the memristive forward/update path compatible with keyed computations.
> 
> ### Kernel backend contract
> * Adapts `_CCInvQuantumLayer` / `FidelityKernel` to g2 sectored raw keys.
> * Accepts `SectoredDistribution` from NoisyG2SLOS internally.
> * Flattens/cleans sectored distributions before returning public tensor kernel outputs.
> * Adds regressions for g2 kernel construction and evaluation.
> 
> ### Deprecations
> * Preserves release/0.4 deprecation behavior.
> * Keeps `no_bunching` deprecation/error behavior.
> * Makes the TorchScript `output_map_func` deprecation warning deterministic instead of relying on PyTorch version-specific warnings.
> 
> ## Documentation
> * Adds noisy simulation documentation in the user guide and quantum expert area.
> * Adds API documentation entries for the new NoisySLOS and sectored distribution surface.
> * Adds diagrams/images for noisy simulation explanations.
> 
> ## How to test / How to run
> 1. Command lines
> 
> ```
> pytest tests/
> ```
> 
> ## Documentation
> * [x]  User docs updated (Sphinx)
> * [x]  Examples / notebooks updated
> * [x]  Docstrings updated
> * [x]  Updated the API
> 
> ## Checklist
> * [x]  PR title includes Jira issue key (e.g., PML-126)
> * [x]  "Related Jira ticket" section includes the Jira issue key (no URL)
> * [x]  Code formatted (ruff format)
> * [x]  Lint passes (ruff)
> * [x]  Static typing passes (mypy) if applicable
> * [x]  Unit tests added/updated (pytest)
> * [x]  Tests pass locally (pytest)
> * [x]  Tests pass on GPU (pytest)
> * [x]  Test coverage not decreased significantly
> * [x]  Docs build locally if affected (sphinx)
> * [x]  With this command: `SPHINXOPTS="-W --keep-going -n" make -C docs clean html` the docs are built without any warning or errors.
> * [x]  New public classes/methods/packages are added in the API following the methodology presented in other files.
> * [x]  Dependencies updated (if needed) and pinned appropriately
> * [x]  PR description explains what changed and how to validate it

since this is a conflict merge ticket and I'd already personally reviewed many of the sub-tickets, it is fine for me once the conflicts are resolved

File: tests/pcvl_pytorch/test_noisy_slos.py [hunk 1]
```diff
@@ -0,0 +1,359 @@
...
+        match="Noisy simulations with source noise currently use ComputationSpace.FOCK. Other computation spaces are not yet supported for noise models.",
+    ):
+        noisy_slos = NoisySLOSComputeGraph(
+            groups, m=5, n_photons=3, computation_space=ComputationSpace.UNBUNCHED
+        )
+        assert noisy_slos.computation_space == ComputationSpace.FOCK
+        noisy_slos = NoisySLOSComputeGraph(
+            groups, m=5, n_photons=3, computation_space=ComputationSpace.UNBUNCHED
+        )
+        assert noisy_slos.computation_space == ComputationSpace.DUAL_RAIL
```
[the remark this answers is not in the document]
Comment (CassNot) [reply tests/pcvl_pytorch/test_noisy_slos.py]: - changed the warning assertion from pytest.raises(UserWarning) to pytest.warns(UserWarning)
- moved the fallback computation_space assertion outside the warning context
- parameterized the test to cover both unsupported spaces (UNBUNCHED and DUAL_RAIL)

File: merlin/algorithms/layer_utils.py [hunk 2]
```diff
@@ -971,3 +1189,251 @@ def normalize_output_key(
...
+            f"Backend '{strategy}' is not supported. "
+            "Only 'slos' is currently available."
+        )
+    return noise
+
+
+def normalize_noise(
+    layer_noise: pcvl.NoiseModel | None,
+    experiment_noise: pcvl.NoiseModel | None,
+) -> pcvl.NoiseModel:
```
[the remark this answers is not in the document]
Comment (CassNot) [reply merlin/algorithms/layer_utils.py:1359]: Done

File: merlin/core/sectored_distribution.py [hunk 3]
```diff
@@ -0,0 +1,461 @@
...
+import torch
+
+from ..utils.combinadics import Combinadics
+from .computation_space import ComputationSpace
+
+
+@dataclass
+class SectorResult:
+    """One photon-number sector of a probability output. If keys are not given, the Combinatics.enumerate_states will be used."""
+
```
[the remark this answers is not in the document]
Comment (CassNot) [reply merlin/core/sectored_distribution.py:14]: Changed twice in this file

File: tests/algorithms/test_layer.py [hunk 4]
```diff
@@ -2113,17 +2136,27 @@ def test_simple_num_photons_modes_and_input_state():
         assert ql.quantum_layer.input_state == pcvl.BasicState(input_state)
 
 
-def test_simple_parameters():
```
Comment (LF-Vigneux) [inline tests/algorithms/test_layer.py:2116]: Why was this test removed

**Review by LF-Vigneux — COMMENTED** [review]
I checked the docs, wrote to you on Teams @CassNot for a wrongly merged comment, no test was removed excpet one. The code it self should be alright if every test passes

File: merlin/core/sectored_distribution.py [hunk 5]
```diff
@@ -0,0 +1,461 @@
...
+        SectorResult(
+            sectors[i],
+            n_modes=dist.sectors[0].n_modes,
+            n_photons=i,
+            keys=tuple(keys_per_sector[i]),
+        )
+        for i in photon_numbers
+    )
+
+    return SectoredDistribution(sectors=sectors_to_return)
```
[the remark this answers is not in the document]
Comment (CassNot) [reply merlin/core/sectored_distribution.py:463]: I fixed the docstring

File: merlin/measurement/process.py [hunk 6]
```diff
@@ -133,6 +134,111 @@ def pcvl_sampler(
             f"Invalid sampling method: {method}. Valid options are: {self.valid_methods}"
         )
 
+    def pcvl_sampler_g2(
+        self, distribution: SectoredDistribution, shots: int, method: str = None
+    ) -> SectoredDistribution:
```
[the remark this answers is not in the document]
Comment (CassNot) [reply merlin/measurement/process.py:139]: Fixed: `pcvl_sampler_g2` now uses `method: str | None = None`

File: docs/source/user_guide/noisy_simulations.rst [hunk 7]
```diff
@@ -0,0 +1,388 @@
...
+
+.. code-block:: text
+
+    2-photon sector: tensor([[0.3431, 0.0000, 0.3431]])
+    3-photon sector: tensor([[0.1066, 0.0355, 0.0355, 0.1066]])
+    4-photon sector: tensor([[0.0110, 0.0000, 0.0074, 0.0000, 0.0110]])
+
+When ``g2 > 0``, the layer returns a
+:class:`~merlin.core.sectored_distribution.SectoredDistribution` because each
+photon-number sector has a different Fock-space size.
```
[the remark this answers is not in the document]
Comment (CassNot) [reply docs/source/user_guide/noisy_simulations.rst]: Fixed (my mistake when merging)

File: tests/core/cloud/test_parammapping.py [hunk 8]
```diff
@@ -116,7 +116,7 @@ def _make_perceval_layer_two_prefixes(prefixes: list[str], counts: list[int], m:
     c = pcvl.Circuit(m)
 
     mode = 0
-    for pref, n in zip(prefixes, counts, strict=True):
+    for pref, n in zip(prefixes, counts):
```
[the remark this answers is not in the document]
Comment (CassNot) [reply tests/core/cloud/test_parammapping.py]: The helper now uses strict zip again

File: merlin/pcvl_pytorch/slos_torchscript.py [hunk 9]
```diff
@@ -1194,34 +1236,42 @@ def load_slos_distribution_computegraph(path):
...
-    if metadata.get("has_output_map_func", False):
-        graph.mapped_indices = saved_data["mapped_indices"]
-        graph.total_mapped_keys = saved_data["total_mapped_keys"]
-        graph.target_indices = saved_data["target_indices"]
+    noise_groups = metadata.get("noise_groups")
+    if noise_groups is not None:
+        graph = NoisySLOSComputeGraph(
+            noise_groups,
+            m,
+            n_photons,
```
[the remark this answers is not in the document]
Comment (CassNot) [reply merlin/pcvl_pytorch/slos_torchscript.py]: Fixed:

- `keep_keys` is now only read in the regular `NoisySLOSComputeGraph` and `SLOSComputeGraph` branch
- the loader detects g2 with `metadata["graph_type"] == "noisy_g2_slos"`
- g2 metadata now reloads as `NoisyG2SLOSComputeGraph`, not `NoisySLOSComputeGraph`
- tests for coverage

File: tests/algorithms/test_layer.py (hunk 4 above)
Comment (CassNot) [reply tests/algorithms/test_layer.py:2116]: It was another miss on my behalf. I restored it !

**Comment by CassNot** [comment]
- solved conflicts
- fixed Copilot comments as per above
- fixed the [0,1] indistinguishability: the error now says:
"Indistinguishability must be in range [0, 1]"
- for the comment on g2 sector weighting counts photons, not sources: I added documentation for the g2 source-slot model so it is documented + added a test for bunched input `[2, 0]` at `g2=0.5` puts all probability in the 4-photon sector, confirming one g2 source slot per intended photon
@claude review

File: merlin/core/process.py [hunk 10]
```diff
@@ -203,31 +295,578 @@ def _setup_computation_graphs(self):
...
+                if probs.ndim == 1:
+                    probs = probs.unsqueeze(0)
+                tensor_probs_per_state.append(probs)
+
+            if not tensor_probs_per_state:
+                raise RuntimeError("No active input states were found.")
+
+            probs_stacked = torch.stack(tensor_probs_per_state, dim=0)
+            selected_weights = weights.to(probs_stacked.dtype)
+            mixed_probs = torch.einsum("is,sbo->ibo", selected_weights, probs_stacked)
```
[the remark this answers is not in the document]
Comment (ben9871) [reply merlin/core/process.py]: > **P2 Badge Keep the parameter batch axis first**
> 
> When source noise is active and the layer has both a batched tensor/StateVector input state and batched angle parameters, this einsum returns probabilities with shape `(state_batch, parameter_batch, outputs)`. The existing superposition path returns `(parameter_batch, state_batch, outputs)` (`_compute_chunked_superposition` uses `bso`), and downstream code treats the first leading dimension as the normal batch dimension. This silently swaps the two batch axes for noisy runs configured via `input_state=StateVector(...)` or `set_input_state(...)` plus batched classical inputs; use `"is,sbo->bio"` (and the matching g2 sector path) to preserve the established axis order.
> 
> Useful? React with 👍 / 👎.

@claude review the latest commit made to fix this issue

**Review by ben9871 — APPROVED** [review]
Applied a small fix to the output batch shapes to be consistent with the rest of merlin contract following codex's review. Louis-Felix doublechecked my fix and we confirmed it was ok. ready to merge!
